README.md
# sso-oidc

[![NPM][npm]][npm-url]
[![Build][build]][build-badge]
[![Coverage][codecov-shield]][codecov]

> Utility to instrument Single Sign-on (SSO) for Node.js and Express.

This project illustrates the basic client/server login flow for Single Page Applications (SPA) using Single Sign-in (SSO) with OpenID Connect (oidc).

The `sso-oidc` module is intended to be used with [Express.js](https://github.com/expressjs/express) with [cookie-session](https://github.com/expressjs/cookie-session) and [body-parser](https://github.com/expressjs/body-parser) middleware. The client implementation is framework agnostic.

## Getting started

### Server-side

```bash
yarn add sso-oidc body-parser cookie-session
```

It is recommended to read environment secrets using [dotenv](https://github.com/motdotla/dotenv) or a similar module.

```bash
yarn add dotenv
```

```js
// server.js
import { json } from 'body-parser';
import session from 'cookie-session';
import express from 'express';
import Strategy from 'sso-oidc';

const app = express()
  .use(json())
  .use(
    session({
      maxAge: 1 * 60 * 1000, // 60 seconds
      name: 'sso-oidc',
      secret: '<SESSION_SECRET>'
    })
  );

const sso = new Strategy({
  redirectUri: '',
  redirectUriLocal: '',
  clientId: '',
  clientSecret: '',
  issuerId: '',
  tokenUrl: '',
  authUrl: '',
  introspectUrl: ''
});

// Returns the silent authorization url.
app.get('/authUrl', sso.getSilentAuthUrl);

// Checks if the current session is valid.
app.get('/check', sso.check);

// Authenticates user using the temporary code returned from silent authorization.
app.post('/callback', sso.token, sso.introspect, (req, res) => {
  res.send({ user_id: req.session.user_id });
});

// Uses a wildcard to authenticate POST requests for a common, protected route.
app.post('/api/*', sso.protect);

// Resets `access_token`, `user_id` but persists session.
app.post('/api/logout', sso.destroy, ({}, res) => res.send({ success: true }));
```

### Client-side

Refer to the [`create-react-app` example](examples/create-react-app) for a basic client login flow using React hooks.

## License

[Apache 2.0](LICENSE)

[npm]: https://img.shields.io/npm/v/sso-oidc.svg?color=blue
[npm-url]: https://npmjs.com/package/sso-oidc
[build]: https://travis-ci.com/ibm/sso-oidc.svg?branch=master
[build-badge]: https://travis-ci.com/ibm/sso-oidc
[codecov]: https://codecov.io/gh/ibm/sso-oidc
[codecov-shield]: https://img.shields.io/codecov/c/github/ibm/sso-oidc.svg


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0](https://github.com/IBM/sso-oidc/releases/tag/v0.1.0) - 2020-02-04

- Initial release


src/Strategy.spec.ts
import { Request, Response, NextFunction } from 'express';
import Strategy from './Strategy';
import fetch from 'node-fetch';
// tslint:disable: no-object-literal-type-assertion

let sso: Strategy;

describe('Strategy', () => {
  beforeAll(() => {
    sso = new Strategy({
      redirectUri: 'redirectUri',
      redirectUriLocal: 'redirectUriLocal',
      clientId: 'clientId',
      clientSecret: 'clientSecret',
      issuerId: 'issuerId',
      tokenUrl: 'tokenUrl',
      authUrl: 'authUrl',
      introspectUrl: 'introspectUrl'
    });
  });

  test('get silentAuthUrl', () => {
    expect(sso.silentAuthUrl).toEqual(
      'authUrl?client_id=clientId&scope=openid&redirect_uri=redirectUri&response_type=code'
    );
  });

  test('check', () => {
    const req = { session: { access_token: undefined, user_id: undefined } } as Request;
    const res = {} as Response;
    res.send = jest.fn().mockReturnValue(res);
    const next: NextFunction = jest.fn();

    sso.check(req, res, next);
    expect(res.send).toBeCalledWith({ valid: false });
    req.session!.user_id = 'user_id';
    sso.check(req, res, next);
    expect(res.send).toBeCalledWith({ valid: true });
  });

  test('protect', () => {
    const req = { session: { access_token: undefined, user_id: undefined } } as Request;
    const res = {} as Response;
    res.status = jest.fn().mockReturnValue(res);
    res.send = jest.fn().mockReturnValue(res);
    const next: NextFunction = jest.fn();

    sso.protect(req, res, next);
    expect(res.status).toBeCalledWith(401);
    expect(res.send).toBeCalledWith({ error: 'Unauthorized user' });

    req.session!.user_id = 'user_id';
    sso.protect(req, res, next);
    expect(next).toBeCalled();
  });

  test('token', async () => {
    // @ts-ignore
    sso.http = jest.fn().mockImplementation(
      () =>
        new Promise(resolve => {
          resolve({ json: () => ({ access_token: 'access_token' }) });
        })
    ) as typeof fetch;
    const req = {
      body: { code: 'code' },
      session: { access_token: undefined, user_id: undefined }
    } as Request;
    const res = {} as Response;
    const next: NextFunction = jest.fn();

    await sso.token(req, res, next);
    expect(req.session!.access_token).toEqual('access_token');
    expect(next).toBeCalled();
  });

  test('introspect', async () => {
    // @ts-ignore
    sso.http = jest.fn().mockImplementation(
      () =>
        new Promise(resolve => {
          resolve({ json: () => ({ sub: 'user_id' }) });
        })
    ) as typeof fetch;
    const req = {
      body: { code: 'code' },
      session: { access_token: 'access_token', user_id: undefined }
    } as Request;
    const res = {} as Response;
    res.status = jest.fn().mockReturnValue(res);
    res.send = jest.fn().mockReturnValue(res);
    const next: NextFunction = jest.fn();

    await sso.introspect(req, res, next);
    expect(req.session!.user_id).toEqual('user_id');
    expect(next).toBeCalled();

    req.session!.access_token = undefined;

    await sso.introspect(req, res, next);
    expect(res.status).toBeCalledWith(401);
    expect(res.send).toBeCalledWith({
      error: {
        message: 'Invalid access_token',
        description: 'An access_token is required to invoke the introspection endpooint'
      }
    });
  });

  test('destroy', () => {
    const req = { session: { access_token: 'access_token', user_id: 'user_id' } } as Request;
    const res = {} as Response;
    const next: NextFunction = jest.fn();

    sso.destroy(req, res, next);
    expect(req.session!.access_token).toEqual(undefined);
    expect(req.session!.user_id).toEqual(undefined);
  });
});


src/Strategy.ts
import { NextFunction, Request, Response } from 'express';
import fetch from 'node-fetch';
import { URLSearchParams } from 'url';

interface ISso {
  access_token?: string;
  user_id?: string;
}

declare global {
  namespace Express {
    // tslint:disable-next-line: interface-name
    export interface Request {
      session?: ISso;
    }
  }
}

interface ISsoOptions {
  redirectUri: string;
  redirectUriLocal?: string;
  clientId: string;
  clientSecret: string;
  issuerId: string;
  tokenUrl: string;
  authUrl: string;
  introspectUrl: string;
}

interface IConfig extends ISsoOptions {
  silentAuthUrl: string;
}

class Strategy {
  private config: IConfig;
  private http = fetch;

  constructor(options: ISsoOptions) {
    this.config = {
      ...options,
      silentAuthUrl: `${options.authUrl}?client_id=${options.clientId}&scope=openid&redirect_uri=${options.redirectUri}&response_type=code`
    };
  }

  public get silentAuthUrl() {
    return this.config.silentAuthUrl;
  }

  public check = (req: Request, res: Response, {}) => {
    res.send({ valid: this.authenticated(req) });
  };

  public protect = (req: Request, res: Response, next: NextFunction) => {
    if (this.authenticated(req)) {
      return next();
    }

    res.status(401).send({ error: 'Unauthorized user' });
  };

  public token = async (req: Request, {}, next: NextFunction) => {
    const body = new URLSearchParams();
    const { clientId, clientSecret, redirectUri, tokenUrl } = this.config;

    body.append('grant_type', 'authorization_code');
    body.append('client_id', clientId);
    body.append('client_secret', clientSecret);
    body.append('redirect_uri', redirectUri);
    body.append('code', req.body.code);

    const response = await this.http(tokenUrl, { method: 'POST', body });
    const data: { access_token: string } = await response.json();

    req.session!.access_token = data.access_token;
    next();
  };

  public introspect = async (req: Request, res: Response, next: NextFunction) => {
    if (req.session!.access_token === undefined) {
      return res.status(401).send({
        error: {
          message: 'Invalid access_token',
          description: 'An access_token is required to invoke the introspection endpooint'
        }
      });
    }

    const body = new URLSearchParams();
    const { clientId, clientSecret, introspectUrl } = this.config;

    body.append('client_id', clientId);
    body.append('client_secret', clientSecret);
    body.append('token', req.session!.access_token!);

    const response = await this.http(introspectUrl, { method: 'POST', body });
    const data: { sub: string } = await response.json();

    req.session!.user_id = data.sub;
    next();
  };

  public destroy(req: Request, {}, next: NextFunction) {
    req.session!.access_token = undefined;
    req.session!.user_id = undefined;
    next();
  }

  private authenticated(req: Request) {
    return req.session && req.session.user_id !== undefined;
  }
}

export default Strategy;


examples/create-react-app/src/App.js
import * as React from 'react';
import * as Router from 'react-router-dom';
import Login from './Login';

function App() {
  const [authenticated, setAuthenticated] = React.useState(false);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        const res = await fetch('/check');
        const data = await res.json();

        setAuthenticated(data.valid);
      } catch (error) {}
    };

    fetchData();
  }, []);

  return (
    <Router.BrowserRouter>
      <>
        {authenticated ? (
          <Router.Switch>
            <Router.Route exact path="/">
              Logged in{' '}
              <button
                type="button"
                onClick={async () => {
                  try {
                    await fetch('/api/logout');
                    setAuthenticated(false);
                  } catch (error) {}
                }}
              >
                Log out
              </button>
            </Router.Route>
            <Router.Redirect to="/" />
          </Router.Switch>
        ) : (
          <Router.Switch>
            <Router.Route exact path="/Login">
              <Login
                onLogin={() => {
                  setAuthenticated(true);
                }}
              />
            </Router.Route>
            <Router.Redirect to="/Login" />
          </Router.Switch>
        )}
      </>
    </Router.BrowserRouter>
  );
}

export default App;


examples/create-react-app/src/Login.js
import * as React from 'react';
import { useLocation, useHistory } from 'react-router-dom';

function Login(props) {
  const history = useHistory();
  const location = useLocation();
  const [authUrl, setAuthUrl] = React.useState(null);

  React.useEffect(() => {
    const code = location.search.split('code=').pop();

    if (code !== undefined && code.length > 0) {
      const fetchData = async () => {
        try {
          const res = await fetch('/code', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ code })
          });
          const data = await res.json();

          if (data.user_id !== undefined) {
            props.onLogin();
            history.push('/');
          } else {
            history.push('/Login');
          }
        } catch (error) {}
      };

      fetchData();
    }
  });

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        const res = await fetch('/authUrl');
        const data = await res.json();

        setAuthUrl(data.url);
      } catch (error) {}
    };

    fetchData();
  }, []);

  return <div>{authUrl != null && <a href={authUrl}>Log in</a>}</div>;
}

export default Login;


src/app/server.ts
import { json } from 'body-parser';
import session from 'cookie-session';
import { config } from 'dotenv';
import express from 'express';
import Strategy from '../Strategy';

config();

const PORT: number = parseInt(process.env.PORT!, 10) || 8080;

const app = express()
  .use(json())
  .use(
    session({
      maxAge: 24 * 60 * 60 * 1000, // 120 seconds
      secret: process.env.CLIENT_ID
    })
  );

const sso = new Strategy({
  redirectUri: process.env.REDIRECT_URI!,
  redirectUriLocal: process.env.REDIRECT_URI_LOCAL!,
  clientId: process.env.CLIENT_ID!,
  clientSecret: process.env.CLIENT_SECRET!,
  issuerId: process.env.ISSUER_ID!,
  tokenUrl: process.env.TOKEN_URL!,
  authUrl: process.env.AUTH_URL!,
  introspectUrl: process.env.INTROSPECT_URL!
});

app.get('/authUrl', ({}, res) => {
  res.send({ url: sso.silentAuthUrl });
});

app.post('/code', sso.token, sso.introspect, (req, res) => {
  res.send({ user_id: req.session?.user_id });
});

app.get('/check', sso.check);

app.post('/callback', sso.token, sso.introspect, (req, res) => {
  res.send({ user_id: req.session!.user_id });
});

app.get('/api/*', sso.protect);

app.get('/api/logout', sso.destroy, ({}, res) => {
  res.send({});
});

app.listen(PORT, () => {
  process.stdout.write(`Listening on port ${PORT}\n`);
});
